### PR TITLE
feat(services) support enabled field TDX-1642

### PR DESCRIPTION
**Pull request by HenriPro**
### Summary

This PR adds support for an enabled boolean field to service entity with a default value of true

When a service is `enabled = false` the proxy behavior will be as if any routes attached to it do not exist. (404) 
Note: this is different from a route with no service, which returns error:  "no Service found with those values" (503)

The primary reason for this feature is to enable easy configuration in Konnect for disabling services to allow for "archiving" and "draft" where configuration can be saved, but for the proxy to act as if it doesn't actually exist. 

EDIT: 1ba6c42 adds filtering to exporting config so when running in hybrid mode, DPs are not sent disabled services and the routes/plugins associated with them. If there is a major issue with the way I implemented that, happy to revert it out of this PR and put it in separate PR. 

As this is core entity change, moving the PR to here instead of ee.

### CODE REVIEW

File: spec/01-unit/01-db/01-schema/11-declarative_config/03-flatten_spec.lua [hunk 1]
```diff
@@ -139,6 +140,7 @@ describe("declarative config: flatten", function()
               tls_verify_depth = null,
               tls_verify = null,
               ca_certificates = null,
+              enabled = true,
```
Comment (mikefero) [inline spec/01-unit/01-db/01-schema/11-declarative_config/03-flatten_spec.lua:143]: Once no longer `required` these `enabled = true` can be removed.

File: spec/02-integration/05-proxy/02-router_spec.lua [hunk 2]
```diff
@@ -235,6 +235,34 @@ for _, strategy in helpers.each_strategy() do
             hosts     = { "serviceless-route-http.test" },
             service   = ngx.null,
           },
+          {
+            paths      = { "/disabled-service1" },
+            protocols  = { "http" },
+            strip_path = false,
+            service    = {
+              path     = "/disabled-service-path/",
+              enabled  = false,
```
Comment (mikefero) [inline spec/02-integration/05-proxy/02-router_spec.lua:244]: If the C* migration suggestion is not added then a test should be added to handle the `nil` case for the `enabled` service configuration.

**Review by mikefero — CHANGES_REQUESTED** [review]
Looks good, have a couple of suggestions.

File: kong/db/schema/entities/services.lua [hunk 3]
```diff
@@ -43,6 +43,7 @@ return {
     { tls_verify         = { type = "boolean", }, },
     { tls_verify_depth   = { type = "integer", default = null, between = { 0, 64 }, }, },
     { ca_certificates    = { type = "array", elements = { type = "string", uuid = true, }, }, },
+    { enabled            = { type = "boolean", required = true, default = true, }, },
```
[the remark this answers is not in the document]
Comment (HenriPro) [reply kong/db/schema/entities/services.lua:46]: I added the default based on @bungle 's suggestion here https://github.com/Kong/kong-ee/pull/2749#discussion_r756844501 

From that it seems that by having both default and required fields allows the dao to fill in the value when it's nil. 
Maybe having the casandra migration is more straightforward anyways.

[the remark this answers is not in the document]
Comment (Mierenga) [reply kong/db/schema/entities/services.lua:46]: Having the C* teardown for defaulting to true will work for the existing rows, but I think new C* rows will not be able to get the `default` of true through `process_auto_fields` unless `required` is also true: (https://github.com/Kong/kong/blob/master/kong/db/schema/init.lua#L1506)

File: spec/01-unit/01-db/01-schema/11-declarative_config/03-flatten_spec.lua (hunk 1 above)
Comment (HenriPro) [reply spec/01-unit/01-db/01-schema/11-declarative_config/03-flatten_spec.lua:143]: I think most of these have to remain as it's checking the result of the declarative config, and the default=true will attach enabled=true.

File: kong/db/migrations/core/014_230_to_270.lua [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+      ]]
+    },
+  
+    cassandra = {
+      up = [[
+        ALTER TABLE services ADD enabled boolean;
+      ]]
+    },
+  }
+
```
Comment (HenriPro) [inline kong/db/migrations/core/014_230_to_270.lua:20]:
```suggestion
  
  
```

Comment (HenriPro) [inline kong/db/migrations/core/014_230_to_270.lua:20]:
```suggestion
 
```

File: kong/db/declarative/init.lua [hunk 5]
```diff
@@ -375,7 +375,7 @@ function declarative.load_into_db(entities, meta)
 end
 
 
-local function export_from_db(emitter, skip_ws)
+local function export_from_db(emitter, skip_ws, skip_disabled_services)
```
Comment (bungle) [inline kong/db/declarative/init.lua:378]:
```suggestion
local function export_from_db(emitter, skip_ws, skip_disabled_entities)
```

I would prefer to call it `skip_disabled_entities`. After this PR we have at least `plugins` and `services` that have `enabled` field.

File: kong/db/declarative/init.lua [hunk 6]
```diff
@@ -404,23 +405,33 @@ local function export_from_db(emitter, skip_ws)
...
         return nil, err
       end
 
-      for _, foreign_name in ipairs(fks) do
-        if type(row[foreign_name]) == "table" then
-          local id = row[foreign_name].id
-          if id ~= nil then
-            row[foreign_name] = id
+      -- do not export disabled services and associated enitties when skip_disabled_services
+      if skip_disabled_services and name == "services" and not row.enabled then
```
Comment (bungle) [inline kong/db/declarative/init.lua:416]:
```suggestion
      if skip_disabled_entities and (name == "services" or name == "plugins") and not row.enabled then
```

Just a suggestion, but I think you get the idea.

I know this is feature creep, but while we are at it, it probably makes sense to deal with plugins too. Plugins are not referenced by any other table, so plugins don't need same treatment as services that related entities should be skipped too.

**Review by bungle — CHANGES_REQUESTED** [review]

File: kong/db/declarative/init.lua (hunk 6 above)
Comment (HenriPro) [reply kong/db/declarative/init.lua]: done

**Review by Mierenga — APPROVED** [review]
Verified migrations and proxy functionality locally with Postgres and C*

**Review by mikefero — APPROVED** [review]
Nice work @HenriPro; any little nickel and dime things can be fixed later. Thanks for adding all the additional features too :partying_face:

**Pull request merged**